loopd/daemon.go

```diff
@@ -119,117 +140,291 @@ func daemon(config *Config, lisCfg *listenerCfg) error {
+		// active.
+		switch d.cfg.Network {
+		case "mainnet":
+			d.cfg.SwapServer = mainnetServer
+		case "testnet":
+			d.cfg.SwapServer = testnetServer
+		default:
+			return errors.New("no swap server address specified")
+		}
+	}
+
+	log.Infof("Swap server address: %v", d.cfg.SwapServer)
+
+	// Create an instance of the loop client library.
+	swapclient, clientCleanup, err := getClient(d.cfg, &d.lnd.LndServices)
+	if err != nil {
+		return err
+	}
+	d.clientCleanup = clientCleanup
+
+	// Both the client RPC server and and the swap server client should
+	// stop on main context cancel. So we create it early and pass it down.
+	d.mainCtx, d.mainCtxCancel = context.WithCancel(context.Background())
+
+	// Now finally fully initialize the swap client RPC server instance.
+	d.swapClientServer = swapClientServer{
+		impl:        swapclient,
+		lnd:         &d.lnd.LndServices,
+		swaps:       make(map[lntypes.Hash]loop.SwapInfo),
+		subscribers: make(map[int]chan<- interface{}),
+		statusChan:  make(chan loop.SwapInfo),
+		mainCtx:     d.mainCtx,
+	}
+
+	// Retrieve all currently existing swaps from the database.
+	swapsList, err := d.impl.FetchSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swapsList {
+		d.swaps[s.SwapHash] = *s
+	}
+
+	// Start the main event handlers, now that we have prepared everything
+	// they need to work.
+	d.startSwapHandlers()
+
+	return nil
+}
 
+// startSwapHandlers starts all goroutines that are tasked with handling swap
+// events.
+func (d *Daemon) startSwapHandlers() {
 	// Start the swap client itself.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Starting swap client")
-		err := swapClient.Run(mainCtx, server.statusChan)
+		err := d.impl.Run(d.mainCtx, d.statusChan)
 		if err != nil {
 			log.Error(err)
+			d.internalErrChan <- err
 		}
 		log.Infof("Swap client stopped")
-
-		log.Infof("Stopping gRPC server")
-		grpcServer.Stop()
-
-		cancel()
 	}()
 
 	// Start a goroutine that broadcasts swap updates to clients.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Waiting for updates")
-		server.processStatusUpdates(mainCtx)
+		d.processStatusUpdates(d.mainCtx)
 	}()
 
-	// Start the grpc server.
-	wg.Add(1)
+	// Last, start our internal error handler. This will return at most one
+	// error on the main error channel to inform the caller that something
+	// went wrong.
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
-		log.Infof("RPC server listening on %s", grpcListener.Addr())
+		defer func() {
+			err := d.Stop()
+			if err != nil {
+				log.Errorf("Error stopping daemon: %v", err)
+			}
+			log.Info("Daemon exited")
+		}()
 
-		if restListener != nil {
-			log.Infof("REST proxy listening on %s", restListener.Addr())
-		}
+		select {
+		case err := <-d.internalErrChan:
+			log.Errorf("Runtime error in daemon, shutting down: "+
+				"%v", err)
+			d.mainErrChan <- err
```

Comment:
This blocks if it isn't read or buffered. I think the usual pattern in `lnd` is to instantiatie the channel in the unit that writes to it, so that it can be created with buffer size `1`.